HARK/ConsumptionSaving/ConsPortfolioModel.py

```diff
@@ -652,42 +652,42 @@ def optimize_share(self):
         Optimization of Share on continuous interval [0,1]
         """
 
-        # For values of aNrm at which the agent wants to put more than 100% into risky asset, constrain them
         FOC_s = self.EndOfPrddvds
-        # Initialize to putting everything in safe asset
-        self.Share_now = np.zeros_like(self.aNrmGrid)
-        self.cNrmAdj_now = np.zeros_like(self.aNrmGrid)
+
+        # For each value of aNrm, find the value of Share such that FOC-Share == 0.
+        crossing = np.logical_and(FOC_s[:, 1:] <= 0.0, FOC_s[:, :-1] >= 0.0)
+        share_idx = np.argmax(crossing, axis=1)
+        a_idx = np.arange(self.aNrmCount)
+        bot_s = self.ShareGrid[share_idx]
+        top_s = self.ShareGrid[share_idx + 1]
+        bot_f = FOC_s[a_idx, share_idx]
+        top_f = FOC_s[a_idx, share_idx + 1]
+        bot_c = self.EndOfPrddvdaNvrs[a_idx, share_idx]
+        top_c = self.EndOfPrddvdaNvrs[a_idx, share_idx + 1]
+        alpha = 1.0 - top_f / (top_f - bot_f)
+
+        self.Share_now = (1.0 - alpha) * bot_s + alpha * top_s
+        self.cNrmAdj_now = (1.0 - alpha) * bot_c + alpha * top_c
+
         # If agent wants to put more than 100% into risky asset, he is constrained
         constrained_top = FOC_s[:, -1] > 0.0
         # Likewise if he wants to put less than 0% into risky asset
         constrained_bot = FOC_s[:, 0] < 0.0
+
+        # For values of aNrm at which the agent wants to put
+        # more than 100% into risky asset, constrain them
         self.Share_now[constrained_top] = 1.0
+        self.Share_now[constrained_bot] = 0.0
```

Comment:
@Mv77 looks like I thought about fixing Share_now for points constrained at the bottom a few years ago but never got around to merging it